### ISSUE TITLE
[confighttp] snappy compress, Encode and NewReader inconsistent behavior

### ISSUE THREAD

**Issue by tomatopunk** [open-telemetry/opentelemetry-collector#10584]
Hello team, I hope you are having a good day.

**Describe the bug**

I am coding on the collect-contrib, about on a new component to receive prometheus_remote_write. 

I have encountered a potential bug related to inconsistent behavior in Snappy compression when using encode -> NewReader and NewBufferedWriter -> decode.

**When we compress or decompress from a stream, or read all of it and then compress or decompress it, we get different results**

**Steps to reproduce**

**What did you expect to see?**

```

package main

import (
	"bytes"
	"fmt"
	"github.com/golang/snappy"
	"io"
)

func main() {
	test1()
	test2()
}

func test1() {
	data := []byte("Hello, World!")
	//encode
	compressed := snappy.Encode(nil, data)

	//decode
	reader := bytes.NewReader(compressed)
	sr := snappy.NewReader(reader)
	sb := new(bytes.Buffer)
	_, err := io.Copy(sb, sr)
	if err != nil {
		println("test1 decode error:", err.Error())
		return
	}

	fmt.Printf("test1 string1: %s, 2: %s", string(data), string(sb.Bytes()))
}

func test2() {
	data := []byte("Hello, World!")

	//encode
	var buf bytes.Buffer
	sw := snappy.NewBufferedWriter(&buf)
	_, err := sw.Write(data)
	if err != nil {
		println("encode error")
		return
	}

	//decode
	byteSlice := buf.Bytes()
	reqBuf, err := snappy.Decode(nil, byteSlice)
	if err != nil {
		println("test2 decode error:", err.Error())
		return
	}

	fmt.Printf("test1 string1: %s, 2: %s", string(data), string(reqBuf))

}

```

**What did you see instead?**

maybe should be correctly code/decode by snappy

**What version did you use?**

main branch

**What config did you use?**

code: https://github.com/open-telemetry/opentelemetry-collector/blob/725e8699bf43c71e1209dccbb444e480bee7bda7/config/confighttp/compression.go#L61

**Environment**

go 1.22

**Additional context**

In the case of Prometheus and OpenTelemetry collection, Prometheus uses encode and decode, while OpenTelemetry uses NewReader and NewBufferedWriter. However, I believe that whether you decode/encode in stream or read all and then encode/decode, the same protocol should yield the same results, and there should be no parsing failures.

code: https://github.com/prometheus/prometheus/blob/c173cd57c921f582586fc725ad51124728757533/cmd/promtool/metrics.go#L104

I would like to know how to address this bug. Should we consider skipping Snappy? If any fixes are required, I am willing to help resolve this issue.

**Comment by BinaryFissionGames** [comment]
This mismatch between the reader/writer and encode/decode is called out in the library: https://pkg.go.dev/github.com/golang/snappy#pkg-overview - it explains there is a "stream" format and a "block" format.

Digging a little deeper, it looks like the stream/framing format should be specified as `x-snappy-framed`, as described here: https://github.com/google/snappy/blob/main/framing_format.txt - I'd expect `snappy` then to be exclusively for the block/non-framed format, like Prometheus uses.

**Comment by jpkrohling** [comment]
Thank you for digging into this, @BinaryFissionGames! I just double-checked, and we are using readers/writers consistently in the Collector code, I believe an OTel Collector as client can compress its data using snappy and a Collector server can read it. I'm closing this issue, but if I'm missing something, let me know and I'll reopen.

reader:
https://github.com/open-telemetry/opentelemetry-collector/blob/6fcebdbf64a8c7eb1c2f17501e93a04a8c19b540/config/confighttp/compression.go#L61-L72

writer:
https://github.com/open-telemetry/opentelemetry-collector/blob/6fcebdbf64a8c7eb1c2f17501e93a04a8c19b540/config/confighttp/compressor.go#L29-L30

**Comment by tomatopunk** [comment]
I believe this issue should be reopened. 

Actually, we are discussing the need for consistency between streaming and framed encryption/decryption. 

For example, stream encryption requires stream decryption, and framed encryption requires framed decryption. However, we are unsure about the correct methods to use for encryption and decryption.

**Comment by tomatopunk** [comment]
In the post, I mentioned to use of reader/write and encode/decode.
Specifically, if we use `newBufferWrited` to encry and use `decode` to decry, we will get an error `snappy is incorrent`

**Comment by tomatopunk** [comment]
As I mentioned in this post, a classic example is the otlp-collector and prometheus remote write.
Prometheus remote write use `snappy.Encode` for compress, while the otlp-collector use `snappy.NewReader`. As a result, I encountered a snappy error, despite both using snappy.

**Comment by tomatopunk** [comment]
so, I suggest either removing snappy decompress in the otlp-collector or replacing the current model with block format.

**Comment by jpkrohling** [comment]
Looking at the documentation for that package, we are using what should be called "x-snappy-framed".

I'm not sure it's that easy for us to change from framed to block right now, as it would mean that clients and servers are not compatible anymore. I'd rather do a phased approach, where, once every couple of versions we take one of the following steps:

1. add "x-snappy-framed" working alongside "snappy" (v0.107.0)
2. deprecate "snappy" (v0.109.0)
3. remove "snappy" (framed) (v0.111.0)
4. add "snappy" (blocked) (v0.113.0)

I believe the risk in breaking clients/servers this way is reduced, and we can then have another snappy support for block format, required by Prometheus remote write. Is it possible to implement framed format in Prometheus remote write? If so, we'd have it working on the next version already, due in a couple of weeks.

**Comment by tomatopunk** [comment]
Perhaps we could allow the receive component to choose whether to automatically decompress.

**Comment by skandragon** [comment]
This is a blocker for prometheus remote write, which I need to implement ASAP.

**Comment by skandragon** [comment]
This is still a blocker for the in-contrib prometheus-remote-write-receiver.  It cannot function today.

**Comment by tomatopunk** [comment]
> This is still a blocker for the in-contrib prometheus-remote-write-receiver. It cannot function today.

Hello @skandragon,

Yes, this is always a question. Perhaps you could send some emails to the code owners, like @codeboten, to get their opinions. The biggest challenge in solving this issue is finding a willing sponsor to support it. 

If I were to implement it, I might choose to extend ServerConfig and modify the newBodyReader method. This could involve either forcibly disabling the Snappy decompression logic or providing a configuration option to specify the decompression type.

https://github.com/open-telemetry/opentelemetry-collector/blob/a8f95d9cef097576a60ba44f120e8dde01e41d62/config/confighttp/compression.go#L186-L193

https://github.com/open-telemetry/opentelemetry-collector/blob/c43ecad7e48522ec1bf4716a975bc240b2e433ce/config/confighttp/confighttp.go#L414

**Comment by skandragon** [comment]
@tomatopunk I went a different route by adding format detection logic to the `confhttp` package.  Hopefully this will be able to make it in eventually, but until then, I have a branch up and working.

I do not think it is correct to keep using "snappy" content-type incorrectly, and this feels like it will unblock that concern, and ease migration to the correct wire format.

**Comment by tomatopunk** [comment]
> [@tomatopunk](https://github.com/tomatopunk) I went a different route by adding format detection logic to the `confhttp` package. Hopefully this will be able to make it in eventually, but until then, I have a branch up and working.
> 
> I do not think it is correct to keep using "snappy" content-type incorrectly, and this feels like it will unblock that concern, and ease migration to the correct wire format.

@skandragon , you are right. We should distinguish between the two types of Snappy instead of using the confusing snappy as the Content-Encoding field. 

However, both Prometheus 1.0 and 2.0 protocols require Content-Encoding: snappy. Changing the protocol would take longer than adjusting the code for compatibility. 

If you want to quickly resolve this issue, you should consider submitting a PR to modify the Prometheus remote write protocol.

https://prometheus.io/docs/specs/prw/remote_write_spec/#definitions

**Comment by skandragon** [comment]
@tomatopunk I think the issue is, `content-type: snappy` means, to everyone _not_ the Otel collector, "block snappy".  That is the real issue.  Prometheus is not going to change, and I do not blame them as they are correct in how they have used that content header and the format of the data they sent.

In any case, the PR I raised seems to have gotten some attention.  Let's see where that lands :)
